### ISSUE TITLE
ClientWebSocket.ConnectAsync shows different behavior when run with a continuation vs not. When auth fails during connect.

### ISSUE THREAD

**Issue by Mossharbor** [dotnet/corefx#36370]
tldr; when auth fails ConnectAsync does not throw an exception when task continuations are used.

I have added this auth code to my service asp.net core service
```csharp
    public class AuthenticationMiddleware
    {
        private readonly RequestDelegate next;

        public AuthenticationMiddleware(RequestDelegate next)
        {
            this.next = next;
        }

        public async Task Invoke(HttpContext context)
        {
            string authHeader = context.Request.Headers["Authorization"];
            if (authHeader != null && authHeader.StartsWith("Basic"))
            {
                // Extract credentials
                string encodedUsernamePassword = authHeader.Substring("Basic ".Length).Trim();
                Encoding encoding = Encoding.GetEncoding("iso-8859-1");
                string usernamePassword = encoding.GetString(Convert.FromBase64String(encodedUsernamePassword));

                int seperatorIndex = usernamePassword.IndexOf(':');

                var username = usernamePassword.Substring(0, seperatorIndex);
                var password = usernamePassword.Substring(seperatorIndex + 1);

                if (username == "testuser" && password == "testpassword")
                {
                    await this.next.Invoke(context).ConfigureAwait(false);
                }
                else
                {
                    context.Response.StatusCode = 401; // Unauthorized
                    return;
                }
            }
            else
            {
                await this.next.Invoke(context).ConfigureAwait(false);
            }
        }
```
.. inside startup.cs
```csharp
public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
            _loggerFactory = loggerFactory;

            app.UseDefaultFiles()
                .UseStaticFiles()
                .UseMiddleware<AuthenticationMiddleware>() // handle the basic authetication test
```

If I call ConnectAsync in a client with improper Authentication like this:
```csharp
string userAndPassword = "testuser:testwrongpassword";
Encoding encoding = Encoding.GetEncoding("iso-8859-1");
var encString = Convert.ToBase64String(encoding.GetBytes(userAndPassword));

CancellationTokenSource source = new CancellationTokenSource();
using (ClientWebSocket cs = new ClientWebSocket())
{
     cs.Options.SetRequestHeader("Authorization", "Basic " + encString);
    cs.ConnectAsync(new Uri(this.botWebSocketUrl), source.Token).Wait();
```

I receive a WebSocketException (as expected).  If I use Task.ContinueWith after the call to connect async like this:

```csharp
cs.ConnectAsync(new Uri(this.botWebSocketUrl), source.Token)
                        .ContinueWith(
                        p =>
                        {
                            Console.WriteLine("Done");
                        },
                        TaskContinuationOptions.OnlyOnRanToCompletion)
                    .ContinueWith(
                       t =>
                       {
                           Console.WriteLine("Exception Thrown");
                       },
                       TaskContinuationOptions.OnlyOnFaulted)
                    .ContinueWith(
                       t =>
                       {
                           Console.WriteLine("Cancelled");
                       },
                       TaskContinuationOptions.OnlyOnCanceled).Wait();
```
I never receive an exception (aka the OnlyOnCanceled option gets hit).

With the [latest changes ](https://github.com/dotnet/corefx/pull/29159/files) I need the exception to see if this is a connection failure or an auth failure.

**Comment by Clockwork-Muse** [comment]
...because each call to `ContinueWith()` is returning a _new_ continuation, so if `ConnectAsync` never completes successfully, the first continuation never even starts (and the rest of your chain remains pending).  You would need to differentiate between states inside one continuation for what you're trying to do.

That, or switch to using an `async` method, and trap exceptions as normal.

**Comment by Mossharbor** [comment]
Gotcha below works.   For some reason I thought the task that was returned by ContinueWith was the original task not a new one.

` .ContinueWith(
                        p =>
                        {
                            if (p.Status == TaskStatus.RanToCompletion)
                            {
                                Console.WriteLine("Done");
                            }
                            else if (p.Status == TaskStatus.Faulted)
                            {
                                Console.WriteLine(p.Exception.Message);
                            }
                            else if (p.Status == TaskStatus.Canceled)
                            {
                                Console.WriteLine("Cancelled");
                            }
                        }).Wait();`
